### PR TITLE
Fix schema parsing for array types in MCP Tool

### LINKED ISSUE #30: MCP Tool Schema Parsing Fails When parameter 'type' is an Array
The tool schema fails to parse when the `type` field is defined as an array, such as `["object", "null"]`, which is a valid construct in [JSON Schema](https://json-schema.org/understanding-json-schema/reference/type.html#type) to allow multiple types for a property.

In our case, the following schema results in a validation error:

```json
{
  "name": "ExampleTool",
  "description": "Tool decription...",
  "strict": false,
  "parameters": {
    "properties": {
      "param1": {
        "description": "Description....",
        "type": "string"
      },
      "kvPairs": {
        "additionalProperties": {
          "type": "string"
        },
        "description": "kvPairs...",
        **// This is where it fail
        "type": [
          "object",
          "null"
        ]
        **
      },
    },
    "required": [
      "param1"
    ],
    "title": "MyTool",
    "type": "object"
  }
}
```

However, the current implementation appears to reject array types, as seen in the following stack trace:

```
Fetched 1 tools from MCP server.
2025-04-10 03:56:07,722 - ERROR - fast_api.py:616 - Error in event_generator: 1 validation error for Schema
type
  Input should be 'TYPE_UNSPECIFIED', 'STRING', 'NUMBER', 'INTEGER', 'BOOLEAN', 'ARRAY' or 'OBJECT' [type=enum, input_value=['object', 'null'], input_type=list]
    For further information visit https://errors.pydantic.dev/2.11/v/enum
Traceback (most recent call last):
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/cli/fast_api.py", line 605, in event_generator
    async for event in runner.run_async(
    ...<8 lines>...
      yield f"data: {sse_event}\n\n"
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/runners.py", line 197, in run_async
    async for event in invocation_context.agent.run_async(invocation_context):
    ...<2 lines>...
      yield event
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/agents/base_agent.py", line 141, in run_async
    async for event in self._run_async_impl(ctx):
      yield event
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/agents/llm_agent.py", line 232, in _run_async_impl
    async for event in self._llm_flow.run_async(ctx):
      self.__maybe_save_output_to_state(event)
      yield event
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/flows/llm_flows/base_llm_flow.py", line 231, in run_async
    async for event in self._run_one_step_async(invocation_context):
      last_event = event
      yield event
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/flows/llm_flows/base_llm_flow.py", line 245, in _run_one_step_async
    async for event in self._preprocess_async(invocation_context, llm_request):
      yield event
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/flows/llm_flows/base_llm_flow.py", line 283, in _preprocess_async
    await tool.process_llm_request(
        tool_context=tool_context, llm_request=llm_request
    )
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/tools/base_tool.py", line 96, in process_llm_request
    if (function_declaration := self._get_declaration()) is None:
                                ~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/tools/mcp_tool/mcp_tool.py", line 94, in _get_declaration
    parameters = to_gemini_schema(schema_dict)
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/tools/openapi_tool/openapi_spec_parser/rest_api_tool.py", line 104, in to_gemini_schema
    k: to_gemini_schema(v) for k, v in value.items()
       ~~~~~~~~~~~~~~~~^^^
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/google/adk/tools/openapi_tool/openapi_spec_parser/rest_api_tool.py", line 133, in to_gemini_schema
    return Schema(**pydantic_schema_data)
  File "/Users/nikhilpandey/Projects/ADKTest/.venv/lib/python3.13/site-packages/pydantic/main.py", line 253, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
pydantic_core._pydantic_core.ValidationError: 1 validation error for Schema
type
  Input should be 'TYPE_UNSPECIFIED', 'STRING', 'NUMBER', 'INTEGER', 'BOOLEAN', 'ARRAY' or 'OBJECT' [type=enum, input_value=['object', 'null'], input_type=list]
    For further information visit https://errors.pydantic.dev/2.11/v/enum
```

### Expected Behavior

ADK should accept array values for the `type` field, such as `["object", "null"]`, and treat them as a valid schema definition.

### Actual Behavior

ADK throws a validation error when encountering a list of types, indicating that only singular types are allowed (e.g., `'STRING'`, `'OBJECT'`, etc.).

### PR DESCRIPTION
**Pull request by ShyamSathish005**
Fixes #30

Update the tool schema parsing to handle array values for the 'type' field.

* Modify `to_gemini_schema` function in `src/google/adk/tools/openapi_tool/openapi_spec_parser/rest_api_tool.py` to handle array values for the 'type' field.
* Add logic to convert array types to the appropriate `Schema` type in `src/google/adk/tools/openapi_tool/openapi_spec_parser/rest_api_tool.py`.
* Add unit tests in `src/google/adk/tests/unittests/tools/openapi_tool/openapi_spec_parser/test_rest_api_tool.py` to verify the `to_gemini_schema` function handles array values for the 'type' field.
* Add test cases for schemas with array types like `["object", "null"]` in `src/google/adk/tests/unittests/tools/openapi_tool/openapi_spec_parser/test_rest_api_tool.py`.

---

For more details, open the [Copilot Workspace session](https://copilot-workspace.githubnext.com/google/adk-python/pull/33?shareId=429270de-5d5e-4019-b664-3f54836a3027).

### CODE REVIEW

File: src/google/adk/tools/openapi_tool/openapi_spec_parser/rest_api_tool.py [hunk 1]
```diff
@@ -130,6 +130,13 @@ def to_gemini_schema(openapi_schema: Optional[Dict[str, Any]] = None) -> Schema:
         # Simple value assignment (int, str, bool, etc.)
         pydantic_schema_data[snake_case_key] = value
 
+  # Handle array types for the 'type' field
+  if "type" in pydantic_schema_data and isinstance(pydantic_schema_data["type"], list):
+    pydantic_schema_data["type"] = [
+        Type(item.upper()) if isinstance(item, str) else item
```
Comment (crystalin) [inline src/google/adk/tools/openapi_tool/openapi_spec_parser/rest_api_tool.py:136]:
```suggestion
        type(item.upper()) if isinstance(item, str) else item
```

**Comment by crystalin** [comment]
This is not working atm

**Comment by LarryStanley** [comment]
Following change work for me

```python
def to_gemini_schema(openapi_schema: Optional[Dict[str, Any]] = None) -> Schema:
  """Converts an OpenAPI schema dictionary to a Gemini Schema object.

  Args:
      openapi_schema: The OpenAPI schema dictionary.

  Returns:
      A Pydantic Schema object.  Returns None if input is None.
      Raises TypeError if input is not a dict.
  """
  if openapi_schema is None:
    return None

  if not isinstance(openapi_schema, dict):
    raise TypeError("openapi_schema must be a dictionary")

  pydantic_schema_data = {}

  # Adding this to force adding a type to an empty dict
  # This avoid "... one_of or any_of must specify a type" error
  if not openapi_schema.get("type"):
    openapi_schema["type"] = "object"

  # Adding this to avoid "properties: should be non-empty for OBJECT type" error
  # See b/385165182
  if openapi_schema.get("type", "") == "object" and not openapi_schema.get(
      "properties"
  ):
    openapi_schema["properties"] = {"dummy_DO_NOT_GENERATE": {"type": "string"}}

  # Special handling for the "priority" property
  if (openapi_schema.get("type", "") == "object" and 
      "properties" in openapi_schema and 
      "priority" in openapi_schema["properties"] and 
      "enum" in openapi_schema["properties"]["priority"]):
    # Ensure priority with enum has type "string"
    openapi_schema["properties"]["priority"]["type"] = "string"
    # Convert enum values to strings if they're numbers
    if isinstance(openapi_schema["properties"]["priority"]["enum"], list):
      openapi_schema["properties"]["priority"]["enum"] = [
          str(item) if isinstance(item, (int, float)) else item
          for item in openapi_schema["properties"]["priority"]["enum"]
      ]

  for key, value in openapi_schema.items():
    snake_case_key = to_snake_case(key)
    # Check if the snake_case_key exists in the Schema model's fields.
    if snake_case_key in Schema.model_fields:
      if snake_case_key in ["title", "default", "format"]:
        # Ignore these fields as Gemini backend doesn't recognize them, and will
        # throw exception if they appear in the schema.
        # Format: properties[expiration].format: only 'enum' and 'date-time' are
        # supported for STRING type
        continue
      if snake_case_key == "properties" and isinstance(value, dict):
        # Special case for priority property with enum
        if "priority" in value and "enum" in value.get("priority", {}):
          value["priority"]["type"] = "string"
          if isinstance(value["priority"]["enum"], list):
            value["priority"]["enum"] = [
                str(item) if isinstance(item, (int, float)) else item
                for item in value["priority"]["enum"]
            ]
        
        pydantic_schema_data[snake_case_key] = {
            k: to_gemini_schema(v) for k, v in value.items()
        }
      elif snake_case_key == "items" and isinstance(value, dict):
        pydantic_schema_data[snake_case_key] = to_gemini_schema(value)
      elif snake_case_key == "any_of" and isinstance(value, list):
        pydantic_schema_data[snake_case_key] = [
            to_gemini_schema(item) for item in value
        ]
      # Important:  Handle cases where the OpenAPI schema might contain lists
      # or other structures that need to be recursively processed.
      elif isinstance(value, list) and snake_case_key not in (
          "enum",
          "required",
          "property_ordering",
      ):
        new_list = []
        for item in value:
          if isinstance(item, dict):
            new_list.append(to_gemini_schema(item))
          else:
            new_list.append(item)
        pydantic_schema_data[snake_case_key] = new_list
      elif isinstance(value, dict) and snake_case_key not in ("properties"):
        # Handle dictionary which is neither properties or items
        pydantic_schema_data[snake_case_key] = to_gemini_schema(value)
      else:
        # Simple value assignment (int, str, bool, etc.)
        pydantic_schema_data[snake_case_key] = value

  # Convert integer enum values to strings
  if "enum" in pydantic_schema_data and isinstance(pydantic_schema_data["enum"], list):
    # If we have an enum, make sure the type is string
    if "type" not in pydantic_schema_data:
      pydantic_schema_data["type"] = "string"
      
    pydantic_schema_data["enum"] = [
        str(item) if isinstance(item, (int, float)) else item 
        for item in pydantic_schema_data["enum"]
    ]

  return Schema(**pydantic_schema_data)
  ```

**Comment by hangfei** [comment]
I believe this is already fixed in the latest wheel. Please give it a try. Feel free to re-open if you still see the issue.
